### PR TITLE
Calcul du dénominateur pour le score

### PR DESCRIPTION
**Pull request by BastienGauthier**
… par commune

### CODE REVIEW

**Comment by BastienGauthier** [comment]
Il reste quelques "TODO" dans le code avant de merge.

File: e_cartomobile/data_transform/compute_bornes.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    df_bornes_communes must contain x_crs_2154 and y_crs_2154 positions of the cities, in CRS 2154 reference.
+
+    Warning : this ponderation is simplified by taking all charging point the commune centers
+
+    """
+    df_b = df_bornes_communes.copy()
+
+    for cluster in df_b.columns:
+        if not (cluster.startswith("x") or cluster.startswith("y")):
+            df_b[cluster + "_completed"] = df_b.apply(
```
Comment (ghost) [inline e_cartomobile/data_transform/compute_bornes.py:85]: Let's use f-strings instead

<details>
<summary><b>Pattern explanation 👈</b></summary>
F-strings have been introduced in Python 3.6 and are considerably faster than `.format` or string addition. For more details, feel free to check <a href='https://stackoverflow.com/a/43123579'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
# msg = "hello " + first_name\nmsg = f"hello {first_name}"
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[string-addition]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: e_cartomobile/data_transform/compute_bornes.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+    df_irve must contain x_crs_2154 and y_crs_2154 positions, in CRS 2154 reference.
+    df_irve must contain power clusters
+
+    """
+    df_b = df_bornes_communes.copy()
+    for cluster in df_b.columns:
+        df_irve_extract = df_irve[df_irve.cluster == cluster]
+        if not (cluster.startswith("x") or cluster.startswith("y")):
+            df_b[cluster + "_completed"] = df_b.apply(
```
Comment (ghost) [inline e_cartomobile/data_transform/compute_bornes.py:112]: Let's use f-strings instead

<details>
<summary><b>Pattern explanation 👈</b></summary>
F-strings have been introduced in Python 3.6 and are considerably faster than `.format` or string addition. For more details, feel free to check <a href='https://stackoverflow.com/a/43123579'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
# msg = "hello " + first_name\nmsg = f"hello {first_name}"
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[string-addition]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: e_cartomobile/data_transform/compute_bornes.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+):
+    return compute_bornes_ponderated(
+        df_bornes_communes_smooth, get_scenario_weight(scenario)
+    )
+
+
+def get_scenario_weight(scenario):
+    # Dummy fonction : to complete
+    if scenario in ["default", "smoothed_uniform"]:
+        return {x: 1 for x in [cluster + "_completed" for cluster in POWER_CLUSTER]}
```
Comment (ghost) [inline e_cartomobile/data_transform/compute_bornes.py:223]: Let's use f-strings instead

<details>
<summary><b>Pattern explanation 👈</b></summary>
F-strings have been introduced in Python 3.6 and are considerably faster than `.format` or string addition. For more details, feel free to check <a href='https://stackoverflow.com/a/43123579'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
# msg = "hello " + first_name\nmsg = f"hello {first_name}"
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[string-addition]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: e_cartomobile/data_transform/compute_bornes.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    score_4_target_commune,
+)
+from e_cartomobile.data_transform.core import get_x_y_from_lat_lon
+
+POWER_CLUSTER = ["Low", "Standard", "Fast", "Very Fast"]
+
+
+# %%
+# Get the bornes count by commuens and power
+def compute_bornes_by_communes(df_irve_power_clean):
```
Comment (ghost) [inline e_cartomobile/data_transform/compute_bornes.py:16]: This section seems to be untested, let's fix that together!

<details>
<summary><b>Pattern explanation 👈</b></summary>
As your program evolves, refactoring and modifications may subtly cause regressions. Adding well-designed tests for your features will help spot them easily and consistently. For more details, feel free to check <a href='https://www.datacamp.com/tutorial/pytest-tutorial-a-hands-on-guide-to-unit-testing'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
import pytest
from my_lib.submodule import MyClass

@pytest.mark.parametrize(
    "input_arg, expected_output",
    [
        [YOUR_INPUT, YOUR_OUTPUT],
    ],)
def test_myclass_custommethod(input_arg, expected_output):
    obj = MyClass()
    assert obj.custommethod(input_arg) == expected_output
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[untested]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: e_cartomobile/data_transform/compute_bornes.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        index="code_commune_INSEE", columns="cluster"
+    )
+    df_bornes_communes.columns = (
+        POWER_CLUSTER  # warning : depends on compute_power_cluster bins
+    )
+
+    return df_bornes_communes
+
+
+def compute_pdc_by_communes(df_irve_power_clean):
```
Comment (ghost) [inline e_cartomobile/data_transform/compute_bornes.py:32]: This section seems to be untested, let's fix that together!

<details>
<summary><b>Pattern explanation 👈</b></summary>
As your program evolves, refactoring and modifications may subtly cause regressions. Adding well-designed tests for your features will help spot them easily and consistently. For more details, feel free to check <a href='https://www.datacamp.com/tutorial/pytest-tutorial-a-hands-on-guide-to-unit-testing'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
import pytest
from my_lib.submodule import MyClass

@pytest.mark.parametrize(
    "input_arg, expected_output",
    [
        [YOUR_INPUT, YOUR_OUTPUT],
    ],)
def test_myclass_custommethod(input_arg, expected_output):
    obj = MyClass()
    assert obj.custommethod(input_arg) == expected_output
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[untested]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: e_cartomobile/data_transform/compute_bornes.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        index="code_commune_INSEE", columns="cluster"
+    )
+    df_bornes_communes.columns = (
+        POWER_CLUSTER  # warning : depends on compute_power_cluster bins
+    )
+
+    return df_bornes_communes
+
+
+def compute_power_cluster(df_irve_power_clean, cluster_labels=POWER_CLUSTER):
```
Comment (ghost) [inline e_cartomobile/data_transform/compute_bornes.py:47]: This section seems to be untested, let's fix that together!

<details>
<summary><b>Pattern explanation 👈</b></summary>
As your program evolves, refactoring and modifications may subtly cause regressions. Adding well-designed tests for your features will help spot them easily and consistently. For more details, feel free to check <a href='https://www.datacamp.com/tutorial/pytest-tutorial-a-hands-on-guide-to-unit-testing'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
import pytest
from my_lib.submodule import MyClass

@pytest.mark.parametrize(
    "input_arg, expected_output",
    [
        [YOUR_INPUT, YOUR_OUTPUT],
    ],)
def test_myclass_custommethod(input_arg, expected_output):
    obj = MyClass()
    assert obj.custommethod(input_arg) == expected_output
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[untested]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: e_cartomobile/data_transform/compute_bornes.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    )
+    df_irve_power_cluster = pd.cut(df_irve_power_clean.puissance_nominale, bins)
+
+    # Set the labels instead of the bins values
+    df_irve_power_cluster = df_irve_power_cluster.map(dict(zip(bins, cluster_labels)))
+
+    return df_irve_power_cluster
+
+
+def clean_power_values(gdf_irve):
```
Comment (ghost) [inline e_cartomobile/data_transform/compute_bornes.py:64]: This section seems to be untested, let's fix that together!

<details>
<summary><b>Pattern explanation 👈</b></summary>
As your program evolves, refactoring and modifications may subtly cause regressions. Adding well-designed tests for your features will help spot them easily and consistently. For more details, feel free to check <a href='https://www.datacamp.com/tutorial/pytest-tutorial-a-hands-on-guide-to-unit-testing'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
import pytest
from my_lib.submodule import MyClass

@pytest.mark.parametrize(
    "input_arg, expected_output",
    [
        [YOUR_INPUT, YOUR_OUTPUT],
    ],)
def test_myclass_custommethod(input_arg, expected_output):
    obj = MyClass()
    assert obj.custommethod(input_arg) == expected_output
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[untested]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

**Review by ghost — COMMENTED** [review]
Thanks for the PR :pray:



<details>
<summary>Other sections</summary>
<li><i>[string-addition] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L87</code></i></li><li><i>[string-addition] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L114</code></i></li><li><i>[string-addition] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L224</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L17</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L32</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L47</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L64</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L74</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L102</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L129</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L149</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L155</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L178</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L213</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L221</code></i></li><li><i>[missing-type-annotations] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L17</code></i></li><li><i>[intermediate-variable] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L18</code></i></li><li><i>[intermediate-variable] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L22</code></i></li><li><i>[intermediate-variable] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L25</code></i></li><li><i>[missing-type-annotations] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L32</code></i></li><li><i>[intermediate-variable] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L33</code></i></li><li><i>[intermediate-variable] detected at <code>e_cartomobile/data_transform/compute_bornes.py:L37</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/core.py:L4</code></i></li><li><i>[untested] detected at <code>e_cartomobile/data_transform/core.py:L12</code></i></li><li><i>[intermediate-variable] detected at <code>e_cartomobile/data_extract/bornes.py:L76</code></i></li><li><i>[intermediate-variable] detected at <code>e_cartomobile/data_extract/bornes.py:L78</code></i></li>
</details>

For a better experience, it's recommended to check the review comments in the tab [Files Changed](https://github.com/dataforgoodfr/batch11_e_cartomobile/pull/33/files?diff=split).
***
_Feedback is a gift! Quack will adjust your review style when you add reactions to a review comment_

**Comment by michaelromagne** [comment]
Seems good to me, the compute_bornes.py module has a lot of functions and could be clearer but that should work well

**Review by michaelromagne — APPROVED** [review]

**Pull request merged**